source/board/microbitv2/nrf52820/power.c

```diff
@@ -19,19 +19,190 @@
+    power_in_wfi = 0;
+
+    gpio_cfg_input(GPIO_REG(RESET_BUTTON), GPIO_IDX(RESET_BUTTON), RESET_BUTTON_PULL);
+
+    // Enable NRF_POWER interrupt for USB removed/detected
+    NRF_POWER->INTENSET = POWER_INTENSET_USBREMOVED_Msk | POWER_INTENSET_USBDETECTED_Msk;
+    NVIC_EnableIRQ(POWER_CLOCK_IRQn);
 }
 
 void power_down()
 {
-    // TODO
+    power_systemoff();
 }
 
 void power_sleep()
 {
-    // TODO
+    power_wfi();
+}
+
+
+static void power_before(bool systemoff)
+{
+    // TODO - wait for USB transmit to computer to finish?
+    /* Wait for debug console output finished. */
+    // while (!(LPUART_STAT_TC_MASK & UART->STAT))
+    // {
+    // }
+
+    uart_uninitialize();
+
+    // TODO - I think maybe uart_uninitialize does this in USART_Uninitialize
+    /* Disable pins to lower current leakage */
+    // PORT_SetPinMux(UART_PORT, PIN_UART_RX_BIT, kPORT_PinDisabledOrAnalog);
+    // PORT_SetPinMux(UART_PORT, PIN_UART_TX_BIT, kPORT_PinDisabledOrAnalog);
+
+    gpio_disable_hid_led();
+    
+    /* Disable I/O pin SWCLK, SWDIO */
+    PORT_OFF();
+
+    if (systemoff) {
+        i2c_deinitialize();
+        // TODO - should i2c_deinitialize call I2Cdrv->Uninitialize()
+        // so I2C1_SCL & I2C1_SDA get defaulted in nrfx_twis_uninit
+
+        // /* PORTC1 is configured as I2C1_SCL */
+        // PORT_SetPinMux(PORTC, 1U, kPORT_PinDisabledOrAnalog);
+        // /* PORTC2 is configured as I2C1_SDA */
+        // PORT_SetPinMux(PORTC, 2U, kPORT_PinDisabledOrAnalog);
+    }
+
+    // Store NRF_GPIOTE state
+    power_gpiote_enabled = NVIC_GetEnableIRQ(GPIOTE_IRQn);
+    power_gpiote_intenset = NRF_GPIOTE->INTENSET;
+    NRF_GPIOTE->INTENCLR = NRF_GPIOTE->INTENSET;
+
+    // Enable IRQ from RESET_BUTTON 
+    gpio_cfg(GPIO_REG(RESET_BUTTON),
+            GPIO_IDX(RESET_BUTTON),
+            NRF_GPIO_PIN_DIR_INPUT,
+            NRF_GPIO_PIN_INPUT_CONNECT,
+            RESET_BUTTON_PULL,
+            NRF_GPIO_PIN_S0S1,
+            RESET_BUTTON_PULL == NRF_GPIO_PIN_PULLUP ? GPIO_PIN_CNF_SENSE_Low : GPIO_PIN_CNF_SENSE_High);
+
+    NRF_GPIOTE->INTENSET = power_gpiote_intenset | ( GPIOTE_INTENSET_PORT_Set << GPIOTE_INTENSET_PORT_Pos);
+    NVIC_EnableIRQ(GPIOTE_IRQn);
+
+    wake_from_usb = 0;
+    wake_from_reset = 0;
+}
+
+static void power_after()
+{
+    // Restore GPIOTE state
+    if (!power_gpiote_enabled) {
+        NVIC_DisableIRQ(GPIOTE_IRQn);
+    }
+    NRF_GPIOTE->INTENSET = power_gpiote_intenset;
+
+    // Disable RESET_BUTTON edge events
+    gpio_cfg_input(GPIO_REG(RESET_BUTTON), GPIO_IDX(RESET_BUTTON), RESET_BUTTON_PULL);
+
+    /* Configure I/O pin SWCLK, SWDIO */
+    PORT_SWD_SETUP();
+    
+    // TODO - done by uart_initialize?
+    // /* re-configure pinmux of disabled pins */
+    // PORT_SetPinMux(UART_PORT, PIN_UART_RX_BIT, (port_mux_t)PIN_UART_RX_MUX_ALT);
+    // PORT_SetPinMux(UART_PORT, PIN_UART_TX_BIT, (port_mux_t)PIN_UART_TX_MUX_ALT);
+
+    uart_initialize();
+    i2c_deinitialize();
+    i2c_initialize();
```

Comment:
I'm guessing this was in the KL27 code as well. Are there cases were we reach this point without having run `i2c_deinitialize()` before?

Reply:
You are right, we can probably reset state without constantly reinitialising the peripheral.
Looks like it was first added in this large commit: https://github.com/mathias-arm/DAPLink/commit/2635ad58ff0ab44194157f2adfffe481541e556f#diff-17dccc6708a0f9d606e9f07f6b4440a2fa9de3f4bb4a2774acba1cadfbe0db6fR337

@martinwork let's leave this for now, could you open an issue in the microbit-foundation/DAPLink-microbit repo to look into this in the future?

@gerargz do you remember if there was an specific issue that required this `i2c_deinitialize(); i2c_initialize();` when it wakes up from sleep?

Reply:
Created https://github.com/microbit-foundation/DAPLink-microbit/issues/128